test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +103,510 @@ UCS_TEST_P(test_ucp_request, test_request_query)
+                             UCP_OP_ATTR_FIELD_CALLBACK,
+        param.user_data = this;
+        param.cb.send   = (ucp_send_nbx_callback_t)ucs_empty_function;
+
+        ucs_queue_head_init(&m_pending);
+        ucp_ep_purge_lanes(ep, ucp_request_purge_enqueue_cb, &m_pending);
+
+        void *request = ucp_ep_flush_internal(ep, 0, &param, NULL, flushed_cb,
+                                              "ep_restart");
+
+        ASSERT_FALSE(UCS_PTR_IS_ERR(request));
+        wait_for_value(&m_completed, (unsigned)1);
+        ucp_request_release(request);
+    }
+
+    void reset_protocol(operation_e op, const std::string &output_proto,
+                        bool sync = false)
+    {
+        ucp_request_param_t param = {0};
+        void *rreq                = NULL;
+        void *sreq                = NULL;
+
+        ucs::fill_random(m_sbuf);
+        ucs::fill_random(m_rbuf);
+
+        switch (op) {
+        case TAG:
+            param.op_attr_mask = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+            if (sync) {
+                sreq = ucp_tag_send_sync_nbx(sender().ep(), m_sbuf.data(),
+                                             m_msg_size, 0, &param);
+            } else {
+                sreq = ucp_tag_send_nbx(sender().ep(), m_sbuf.data(),
+                                        m_msg_size, 0, &param);
+            }
+
+            ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+            rreq = ucp_tag_recv_nbx(receiver().worker(), m_rbuf.data(),
+                                    m_msg_size, 0, 0, &param);
+            ASSERT_TRUE(UCS_PTR_IS_PTR(rreq));
+            break;
+
+        case RMA_GET:
+            param.op_attr_mask = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+            sreq = ucp_get_nbx(sender().ep(), m_rbuf.data(), m_msg_size,
+                               (uint64_t)m_sbuf.data(), m_rkey, &param);
+            break;
+
+        case RMA_PUT:
+            sreq = ucp_put_nbx(sender().ep(), m_rbuf.data(), m_msg_size,
+                               (uint64_t)m_sbuf.data(), m_rkey, &param);
+            break;
+
+        case STREAM:
+            param.op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE;
+            param.datatype     = ucp_dt_make_contig(1);
+            sreq = ucp_stream_send_nbx(sender().ep(), m_sbuf.data(), m_msg_size,
+                                       &param);
+            break;
+
+        case AM:
+            sreq = send_am();
+            break;
+        }
+
+        ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+        wait_and_restart(sreq, rreq, output_proto);
+        wait_receive(op);
+
+        request_wait(sreq);
+        request_wait(rreq);
+        flush_ep(sender());
+
+        EXPECT_EQ(m_sbuf, m_rbuf);
+    }
+
+    virtual void
+    wait_and_restart(void *sreq, void *rreq, const std::string &expected_proto)
+    {
+        ucp_request_t *req                 = (ucp_request_t*)sreq - 1;
+        const ucp_datatype_iter_t *dt_iter = &req->send.state.dt_iter;
+
+        while (dt_iter->offset == 0) {
```

Comment:
`wait_for_value`

Reply:
I don't wait for a specific value, I wait for any value other then zero

Reply:
ok, so maybe add something like `wait_for_changed_value(addr, value)`? anyway need to add timeout to all loops like this to avoid hanged test in CI